polygeist/tools/cgeist/Lib/Attributes.cc

```diff
@@ -0,0 +1,329 @@
+  case llvm::Attribute::AttrKind::NoCapture:
+    addPassThroughAttribute(createStringAttr("nocapture"));
+    break;
+  /// Function does not deallocate memory.
+  case llvm::Attribute::AttrKind::NoFree:
+    addPassThroughAttribute(createStringAttr("nofree"));
+    break;
+  /// Function is never inlined.
+  case llvm::Attribute::AttrKind::NoInline:
+    addPassThroughAttribute(createStringAttr("noinline"));
+    break;
+  /// Pointer is known to be not null.
+  case llvm::Attribute::AttrKind::NonNull:
+    addPassThroughAttribute(createStringAttr("nonnull"));
+    break;
+  /// Function does not recurse.
+  case llvm::Attribute::AttrKind::NoRecurse:
+    addPassThroughAttribute(createStringAttr("norecurse"));
+    break;
+  /// Function does not return.
+  case llvm::Attribute::AttrKind::NoReturn:
+    addPassThroughAttribute(createStringAttr("noreturn"));
+    break;
+  /// Parameter or return value may not contain uninitialized or poison bits.
+  case llvm::Attribute::AttrKind::NoUndef:
+    addPassThroughAttribute(createStringAttr("noundef"));
+    break;
+  /// Function doesn't unwind stack.
+  case llvm::Attribute::AttrKind::NoUnwind:
+    addPassThroughAttribute(createStringAttr("nounwind"));
+    break;
+  /// Function does not access memory.
+  case llvm::Attribute::AttrKind::ReadNone:
+    addPassThroughAttribute(createStringAttr("readnone"));
+    break;
+  /// Function only reads from memory.
+  case llvm::Attribute::AttrKind::ReadOnly:
+    addPassThroughAttribute(createStringAttr("readonly"));
+    break;
+  /// Hidden pointer to structure to return.
+  case llvm::Attribute::AttrKind::StructRet:
+    addPassThroughAttribute(createStringAttr("sret"));
+    break;
+  /// Function always comes back to callsite.
+  case llvm::Attribute::AttrKind::WillReturn:
+    addPassThroughAttribute(createStringAttr("willreturn"));
+    break;
+  /// Function only writes to memory.
+  case llvm::Attribute::AttrKind::WriteOnly:
+    addPassThroughAttribute(createStringAttr("writeonly"));
+    break;
+  default:
+    llvm_unreachable("Unexpected attribute kind");
+  }
+
+  return *this;
+}
+
+AttrBuilder &AttrBuilder::addPassThroughAttribute(mlir::Attribute attr) {
+  NamedAttribute passThrough = getOrCreatePassThroughAttr();
+  assert(passThrough.getValue().isa<ArrayAttr>() &&
+         "PassThrough attribute should have an ArrayAttr as value");
+
+  LLVM_DEBUG(llvm::dbgs() << "Adding attribute " << attr
+                          << " to 'passthrough'.\n");
+  std::vector<mlir::Attribute> vec =
+      passThrough.getValue().cast<ArrayAttr>().getValue().vec();
+  vec.push_back(attr);
+  passThrough.setValue(ArrayAttr::get(&ctx, vec));
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "  passthrough: ( ";
+    for (auto item : vec)
+      llvm::dbgs() << item << " ";
+    llvm::dbgs() << ")\n";
+  });
+
+  return addAttribute(passThrough);
+}
+
+bool AttrBuilder::contains(StringRef attrName) const {
+  return getAttr(attrName).hasValue();
```

Comment:
Is there any way to query about attributes contained within the "passthrough" attribute?